### DISCUSSION TITLE
Elsa V2: RabbitMQ (Rebus) Queue Balancing / How can we finetune?

### DISCUSSION THREAD

**Discussion by dao74** [elsa-workflows/elsa-core#4332] [Design & tech]
Hi all,

We did some stress test (Bulk operations) on our environment which is a system with 5 nodes.

In result, we saw that the queues are not well balanced; as you can see in the SS below one of the nodes has above 50k the others are around 5k.

![image](https://github.com/elsa-workflows/elsa-core/assets/128688938/8a2400d9-b543-4093-a0cd-2d5a43f0b524)

On other side, we did not understand why each instance has their own queue - because one node has stopped working and the queue was not consumed anymore. Can we change to bind all nodes to a single queue?

How can we finetune to have a better balancing of the queue messages?

Is it possible that huge queues will stop / prevent new workflows to be created?

**Comment by sfmskywalker** [comment]
Hi @dao74,

Thanks for your questions! Here's a breakdown of the answers:

1. **Why Each Instance Has Its Own Input Queue**: Each instance has its own input queue for workflow management events to ensure that every node in your cluster receives notifications. If we used a single queue, only one node would consume the message, potentially leaving other nodes with outdated information. These events play a crucial role, such as in invalidating local caches.

2. **Event Distribution Across Queues**: Ideally, each queue should receive an equal number of events. However, if a node goes offline temporarily, its queue will continue to accumulate messages. This can result in varying message counts across different queues.

3. **Queue Size Limit Concerns**: You're right; if a queue reaches its size limit, the message bus client might start throwing exceptions. This could subsequently affect the handling of workflows within a node.

I hope this clarifies things! If you have any more questions or need further details, feel free to ask.
